dataplane/standalone/apigen/xml.go

```diff
@@ -0,0 +1,195 @@
+}
+
+// SectionDef contains a list of members.
+type SectionDef struct {
+	MemberDef []MemberDef `xml:"memberdef"`
+}
+
+// MemberDef is the definition of a single type.
+type MemberDef struct {
+	Name       string      `xml:"name"`
+	EnumValues []EnumValue `xml:"enumvalue"`
+	Kind       string      `xml:"kind,attr"`
+}
+
+// EnumValue is a single values in a enum.
+type EnumValue struct {
+	Name                string              `xml:"name"`
+	Initializer         string              `xml:"initializer"`
+	DetailedDescription DetailedDescription `xml:"detaileddescription"`
+}
+
+// DetailedDescription contains extra information about an enum value.
+type DetailedDescription struct {
+	Para Para `xml:"para"`
+}
+
+// Para is a generic paragraph.
+type Para struct {
+	SinpleSect []SinpleSect `xml:"simplesect"`
+}
+
+// SinpleSect contains a description of an element.
+type SinpleSect struct {
+	Para string `xml:"para"`
+}
+
+const xmlPath = "dataplane/standalone/apigen/xml"
+
+func parseXml() (*xmlInfo, error) {
+	i := &xmlInfo{
+		attrs: make(map[string]attrInfo),
+		enums: make(map[string][]string),
+	}
+	files, err := os.ReadDir(xmlPath)
+	if err != nil {
+		return nil, err
+	}
+	for _, file := range files {
+		if err := parseXmlFile(filepath.Join(xmlPath, file.Name()), i); err != nil {
+			return nil, err
+		}
+	}
+	return i, nil
+}
+
+var typeNameExpr = regexp.MustCompile("sai_(.*)_attr.*")
+
+func handleEnumAttr(enum MemberDef) attrInfo {
+	info := attrInfo{}
+	trimStr := strings.TrimSuffix(strings.TrimPrefix(enum.Name, "_"), "_t") + "_"
+	for _, value := range enum.EnumValues {
+		var canCreate, canRead, canSet bool
+		var saiType string
+		for _, details := range value.DetailedDescription.Para.SinpleSect {
+			annotation := strings.TrimSpace(details.Para)
+			switch {
+			case strings.HasPrefix(annotation, "@@type"):
+				saiType = strings.TrimSpace(strings.TrimPrefix(annotation, "@@type"))
+			case strings.HasPrefix(annotation, "@@flags"):
+				switch {
+				case strings.Contains(annotation, "CREATE_ONLY"):
+					canCreate = true
+					canRead = true
+				case strings.Contains(annotation, "CREATE_AND_SET"):
+					canCreate = true
+					canRead = true
+					canSet = true
+				case strings.Contains(annotation, "READ_ONLY"):
+					canRead = true
+				}
+			}
+		}
+		if !canCreate && !canRead && !canSet {
+			continue
+		}
+		atn := attrTypeName{
+			EnumName:   value.Name,
+			MemberName: strings.TrimPrefix(strings.ToLower(value.Name), trimStr),
+			SaiType:    saiType,
+		}
+		if canCreate {
+			info.createFields = append(info.createFields, atn)
+		}
+		if canRead {
+			info.readFields = append(info.readFields, atn)
+		}
+		if canSet {
+			info.setFields = append(info.setFields, atn)
+		}
+	}
+	return info
+}
+
+func handleEnum(enum MemberDef) []string {
+	res := []string{}
+	for _, value := range enum.EnumValues {
+		res = append(res, value.Name)
+	}
+	return res
+}
+
+func parseXmlFile(file string, xmlInfo *xmlInfo) error {
```

Comment:
Is there no good XML file parsers out there? https://pkg.go.dev/encoding/xml ?

Reply:
what do you mean i am using encoding/xml

Reply:
Oh oops, sorry about that, just seemed more complicated than I thought it would look. Let me take a more thorough look this time.

Reply:
i've added more comments to clarify

Reply:
LGTM